File: AppBox/Common/ServiceIntegrations/SlackClient/SlackClient.m
//
//  SlackClient.m
//  AppBox
//
//  Created by Vineet Choudhary on 10/04/17.
//  Copyright © 2017 Developer Insider. All rights reserved.
//

#import "SlackClient.h"

@implementation SlackClient

+ (void)sendMessage:(IPAUploadInfo *)ipaUploadInfo
			webhook:(NSString *)webhook
			message:(NSString *)message
		 completion:(void (^) (BOOL success))completion {
	// Validate webhook URL
	if (![Common isValidWebhookURL:webhook]) {
		DDLogError(@"Slack Error - Invalid webhook URL: %@", webhook);
		completion(NO);
		return;
	}

	//set slack channel url and image
	NSString *slackImage = abSlackImage;

	//set slack message
	NSString *finalMessage;
	if (message.length > 0) {
		finalMessage = [MailHandler parseMessage:message forIPAUploadInfo:ipaUploadInfo];
	} else {
		finalMessage = [NSString stringWithFormat:@"%@ - %@ (%@) link - %@", ipaUploadInfo.name, ipaUploadInfo.version, ipaUploadInfo.build, ipaUploadInfo.appShortShareableURL];
	}

	//create parameters dictionary
	NSDictionary *parameters = @{
		@"username": @"AppBox",
		@"icon_url": slackImage,
		@"text": finalMessage
	};

	//send message
	[NetworkHandler requestWithURL:webhook withParameters:parameters andRequestType:RequestPOST andCompletetion:^(id responseObj, NSInteger httpStatus, NSError *error) {
		if (responseObj && error == nil) {
			DDLogInfo(@"Slack Response - %@", responseObj);
			completion(YES);
		} else if (error) {
			DDLogInfo(@"Slack Error - %@", error.localizedDescription);
			completion(NO);
		} else {
			DDLogInfo(@"Slack Error - Unknown Error");
			completion(NO);
		}
	}];
}

@end

File: AppBox/Common/ServiceIntegrations/SlackClient/SlackClient.swift
//
//  SlackClient.swift
//  AppBox

import Foundation
import AppBoxCore
import os

public final class SlackClient: NSObject {

    private static let log = Logger(subsystem: "com.developerinsider.AppBox", category: "SlackClient")

    public class func sendMessage(_ ipaUploadInfo: IPAUploadInfo, webhook: String?,
                                  completion: @escaping (Bool) -> Void) {
        guard WebhookURL.isValid(webhook), let webhook else {
            log.error("Slack Error - Invalid webhook URL: \(webhook ?? "", privacy: .public)")
            completion(false)
            return
        }

        let finalMessage = BuildNotificationMessage.text(
			name: ipaUploadInfo.name ?? "",
			version: ipaUploadInfo.version ?? "",
			build: ipaUploadInfo.build ?? "",
			installURL: ipaUploadInfo.appShortShareableURL?.absoluteString ?? "")

        Task {
            do {
                try await AppServices.serviceClient.sendNotification(service: .slack, webhookURL: webhook, text: finalMessage)
                log.info("Slack message sent")
                await MainActor.run { completion(true) }
            } catch {
                log.info("Slack Error - \(error.localizedDescription, privacy: .public)")
                await MainActor.run { completion(false) }
            }
        }
    }
}
